File: ServiceAreaSample/Objective-C/ServiceAreaSample/ServiceAreaViewController.m
//
//  ViewController.m
//  ServiceAreaSample
//
//  Copyright © 2559 Globtech. All rights reserved.
//

#import "ServiceAreaViewController.h"
#define REFERRER @""

@interface ServiceAreaViewController ()

@end

@implementation ServiceAreaViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    [self initializeMap];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (void)mapView:(AGSMapView *)mapView didClickAtPoint:(CGPoint)screen mapPoint:(AGSPoint *)mappoint features:(NSDictionary *)features {
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"Pin Location"
                                                       style:UIAlertActionStyleDefault
                                                     handler:^(UIAlertAction * _Nonnull action) {
                                                         
                                                         [_graphicLayer removeAllGraphics];
                                                         [_serviceAreaLayer removeAllGraphics];
                                                         AGSGraphic *graphic = [AGSGraphic graphicWithGeometry:mappoint
                                                                                                        symbol:[AGSPictureMarkerSymbol pictureMarkerSymbolWithImageNamed:@"pin_map"]
                                                                                                    attributes:nil];
                                                         
                                                         [_graphicLayer addGraphic:graphic];
                                                         
                                                         AGSPoint *point = [AGSPoint pointFromDecimalDegreesString:[mappoint decimalDegreesStringWithNumDigits:7]
                                                                                              withSpatialReference:[AGSSpatialReference wgs84SpatialReference]];
                                                         
                                                         NTLocation *location = [[NTLocation alloc] initWithName:@"facility" lat:point.y lon:point.x];
                                                         
                                                         NSError *error = nil;
                                                         
                                                         NTServiceAreaParameter *param = [[NTServiceAreaParameter alloc] initWithFacilities:@[location]
                                                                                                                                     breaks:@[@1,@3,@5]];
                                                                                          
                                                
                                                         NTServiceAreaResultSet *resultSet = [NTServiceAreaService execute:param error:&error];
                                                                                          
                                                         
                                                         if (!error && resultSet.results.count == 3) {
                                                                                              
                                                             
                                                             NTServiceAreaResult *result = resultSet.results[2];
                                                             
                                                             AGSPolygon *polygon = [[AGSPolygon alloc] initWithJSON:[result getShape] spatialReference:[AGSSpatialReference wgs84SpatialReference]];
                                                             
                                                             AGSGeometry *geometry = [[AGSGeometryEngine defaultGeometryEngine] projectGeometry:polygon toSpatialReference:[AGSSpatialReference webMercatorSpatialReference]];
                                                                                              
                                                             
                                                             AGSSimpleFillSymbol *fillSymbol = [AGSSimpleFillSymbol simpleFillSymbolWithColor:[UIColor redColor] outlineColor:[UIColor redColor]];
                                                                                              
                                                             
                                                             AGSGraphic *grpahic = [AGSGraphic graphicWithGeometry:geometry symbol:fillSymbol attributes:nil];
                                                                                              
                                                             
                                                             [_serviceAreaLayer addGraphic:graphic];
                                                                                              
                                                             
                                                             result = resultSet.results[1];
                                                             
                                                             polygon = [[AGSPolygon alloc] initWithJSON:[result getShape] spatialReference:[AGSSpatialReference wgs84SpatialReference]];
                                                             
                                                             geometry = [[AGSGeometryEngine defaultGeometryEngine] projectGeometry:polygon toSpatialReference:[AGSSpatialReference webMercatorSpatialReference]];
                                                                                              
                                                             
                                                             fillSymbol = [AGSSimpleFillSymbol simpleFillSymbolWithColor:[UIColor yellowColor] outlineColor:[UIColor yellowColor]];
                                                                                              
                                                             
                                                             grpahic = [AGSGraphic graphicWithGeometry:geometry symbol:fillSymbol attributes:nil];
                                                                                              
                                                             
                                                             [_serviceAreaLayer addGraphic:graphic];
                                                                                              
                                                             
                                                             result = resultSet.results[0];
                                                             
                                                             polygon = [[AGSPolygon alloc] initWithJSON:[result getShape] spatialReference:[AGSSpatialReference wgs84SpatialReference]];
                                                             
                                                             geometry = [[AGSGeometryEngine defaultGeometryEngine] projectGeometry:polygon toSpatialReference:[AGSSpatialReference webMercatorSpatialReference]];
                                                                                              
                                                             
                                                             fillSymbol = [AGSSimpleFillSymbol simpleFillSymbolWithColor:[UIColor greenColor] outlineColor:[UIColor greenColor]];
                                                                                              
                                                             
                                                             grpahic = [AGSGraphic graphicWithGeometry:geometry symbol:fillSymbol attributes:nil];
                                                                                              
                                                             
                                                             [_serviceAreaLayer addGraphic:graphic];
                                                                                              
                                                             
                                                             [_mapView zoomToGeometry:geometry withPadding:100 animated:true];
                                                             
                                                         }
                                                                                          
                                                         
                                                     }];
    
    [alertController addAction:okAction];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    
    [self presentViewController:alertController animated:true completion:nil];
    
}

- (void)mapViewDidLoad:(AGSMapView *)mapView {
    
    [mapView.locationDisplay startDataSource];

    AGSEnvelope *env = [AGSEnvelope envelopeWithXmin:10458701.000000
                                                ymin:542977.875000
                                                xmax:11986879.000000
                                                ymax:2498290.000000
                                    spatialReference:[AGSSpatialReference webMercatorSpatialReference]];
    
    [mapView zoomToEnvelope:env animated:true];
    
    _graphicLayer = [[AGSGraphicsLayer alloc] init];
    _serviceAreaLayer = [[AGSGraphicsLayer alloc] init];
    
    [mapView addMapLayer:_serviceAreaLayer withName:@"ServiceGraphicLayer"];
    [mapView addMapLayer:_graphicLayer withName:@"GraphicLayer"];
    
    
}



- (void)initializeMap {
    _mapView.layerDelegate = self;
    _mapView.touchDelegate = self;
    NSError *error = nil;
    // Get Map Permission
    NTMapPermissionResultSet *resultSet = [NTMapPermissionService executeAndReturnError:&error];
    
    if (!error) {
        
        // Get Street map HD (service id: 2)
        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"serviceId == 2"];
        NSArray *filtered = [resultSet.results filteredArrayUsingPredicate:predicate];
        
        if (filtered.count > 0) {
            NTMapPermissionResult *mapPermission = filtered.firstObject;
            
            NSURL *url = [NSURL URLWithString:mapPermission.serviceUrl_L];
            AGSCredential *cred = [[AGSCredential alloc] initWithToken:mapPermission.serviceToken_L
                                                               referer:REFERRER];
            
            AGSTiledMapServiceLayer *tiled = [[AGSTiledMapServiceLayer alloc] initWithURL:url
                                                                               credential:cred];
            [_mapView addMapLayer:tiled withName:mapPermission.serviceName];
            
        }
    }
    else {
        NSLog(@"load map permission error: %@", error.localizedDescription);
    }
}

@end

File: ServiceAreaSample/Swift/ServiceAreaSample/ServiceAreaViewController.swift
//
//  ViewController.swift
//  ServiceAreaSample
//
//  Copyright © 2559 Globtech. All rights reserved.
//

import UIKit
import ArcGIS
import NOSTRASDK

class ServiceAreaViewController: UIViewController, AGSMapViewLayerDelegate, AGSMapViewTouchDelegate, AGSCalloutDelegate, AGSLayerDelegate {
    
    let referrer = ""
    
    @IBOutlet weak var mapView: AGSMapView!
    
    let graphicLayer = AGSGraphicsLayer();
    let serviceAreaLayer = AGSGraphicsLayer();
    
    let redAreaColor = UIColor(red: 255.0/255.0, green: 0, blue: 0, alpha: 0.5);
    let yellowAreaColor = UIColor(red: 255.0/255.0, green: 255.0/255.0, blue: 0, alpha: 0.5);
    let greenAreaColor = UIColor(red: 0, green: 255.0/255.0, blue: 0, alpha: 0.5);
    var idenResult: NTIdentifyResult?;
    
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        self.initializeMap();
        
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    @IBAction func btnCurrentLocation_Clicked(_ btn: UIButton) {
        
        btn.isSelected = !btn.isSelected;
        mapView.locationDisplay.autoPanMode = btn.isSelected ? .default : .off;
        
    }
    
    
    //MARK: Touch Delegate
    func mapView(_ mapView: AGSMapView!, didClickAt screen: CGPoint, mapPoint mappoint: AGSPoint!, features: [AnyHashable: Any]!) {
        
        
        let alertController = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        

        alertController.addAction(UIAlertAction(title: "Pin Location", style: .default) {
            (alert) in
            

            self.graphicLayer.removeAllGraphics();
            self.serviceAreaLayer.removeAllGraphics();
            
            let graphic = AGSGraphic(geometry: mappoint,
                                     symbol: AGSPictureMarkerSymbol(imageNamed: "pin_map"),
                                     attributes: nil);
            
            self.graphicLayer.addGraphic(graphic);
            
            let point = AGSPoint(fromDecimalDegreesString: mappoint.decimalDegreesString(withNumDigits: 7),
                with: AGSSpatialReference.wgs84());
            
            
            
            let param = NTServiceAreaParameter(facilities: [NTLocation(name: "facility", lat: (point?.y)!, lon: (point?.x)!)], breaks: [1, 3, 5])
            
            do {
                let result = try NTServiceAreaService.execute(param)
                
                var area: NTServiceAreaResult! = nil;
                
                if  result.results.count == 3 {
                    
                    area = result.results[2];
                    var polygon = AGSPolygon(json: area.getShape(), spatialReference: AGSSpatialReference.wgs84())
                    var geometry = AGSGeometryEngine.default().projectGeometry(polygon, to: AGSSpatialReference.webMercator())
                    var symbol = AGSSimpleFillSymbol(color: self.redAreaColor, outlineColor: self.redAreaColor);
                    var graphic = AGSGraphic(geometry: geometry, symbol: symbol, attributes: nil);
                    
                    self.serviceAreaLayer.addGraphic(graphic);
                    
                    area = result.results[1];
                    polygon = AGSPolygon(json: area.getShape(), spatialReference: AGSSpatialReference.wgs84())
                    geometry = AGSGeometryEngine.default().projectGeometry(polygon, to: AGSSpatialReference.webMercator())
                    symbol = AGSSimpleFillSymbol(color: self.yellowAreaColor, outlineColor: self.yellowAreaColor);
                    graphic = AGSGraphic(geometry: geometry, symbol: symbol, attributes: nil);
                    
                    self.serviceAreaLayer.addGraphic(graphic);
                    
                    area = result.results[0];
                    polygon = AGSPolygon(json: area.getShape(), spatialReference: AGSSpatialReference.wgs84())
                    geometry = AGSGeometryEngine.default().projectGeometry(polygon, to: AGSSpatialReference.webMercator())
                    symbol = AGSSimpleFillSymbol(color: self.greenAreaColor, outlineColor: self.greenAreaColor);
                    graphic = AGSGraphic(geometry: geometry, symbol: symbol, attributes: nil);
                    
                    self.serviceAreaLayer.addGraphic(graphic);
                    
                    mapView.zoom(to: geometry, withPadding: 10, animated: true);
                }
                
                
                
                
            }
            catch let error as NSError {
                print("error: \(error.description)");
            }
            
            
            });
        
        alertController.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        
        self.present(alertController, animated: true, completion: nil);

    }
    
    
    //MARK: Layer delegate
    func mapViewDidLoad(_ mapView: AGSMapView!) {
        mapView.locationDisplay.startDataSource()
        
        let env = AGSEnvelope(xmin: 10458701.000000, ymin: 542977.875000,
                              xmax: 11986879.000000, ymax: 2498290.000000,
                              spatialReference: AGSSpatialReference.webMercator());
        mapView.zoom(to: env, animated: true);
        
        mapView.addMapLayer(serviceAreaLayer);
        mapView.addMapLayer(graphicLayer);
        
        
    }
    
    func initializeMap() {
        
        mapView.layerDelegate = self;
        mapView.touchDelegate = self;
        mapView.callout.delegate = self;
        
        do {
            
            // Get map permisson.
            let resultSet = try NTMapPermissionService.execute();
            
            // Get Street map HD (service id: 2)
            if resultSet.results != nil && resultSet.results.count > 0 {
                let filtered = resultSet.results.filter({ (mp) -> Bool in
                    return mp.serviceId == 2;
                })
                
                if filtered.count > 0 {
                    let mapPermisson = filtered.first;
                    
                    
                    let url = URL(string: mapPermisson!.serviceUrl_L);
                    let cred = AGSCredential(token: mapPermisson?.serviceToken_L, referer: referrer);
                    let tiledLayer = AGSTiledMapServiceLayer(url: url, credential: cred)
                    tiledLayer?.delegate = self;
                    
                    mapView.addMapLayer(tiledLayer, withName: mapPermisson!.serviceName);
                }
                
            }
        }
        catch let error as NSError {
            print("error: \(error)");
        }
    }
    
}
